File: src/services/fuse-monitor.service.js
/**
 * Fuse Monitor Service
 * Coordinates GPIO, I2C, and CSV logging for fuse monitoring system
 */

const ADS1115Service = require("./ads1115.service");
const GPIOService = require("./gpio.service");
const CSVLogger = require("./csv.logger");

class FuseMonitorService {
	constructor() {
		this.initialized = false;
		this.monitoring = false;
		this.monitorInterval = null;
		this.totalFuses = 64; // 4 muxes × 16 channels each
		this.totalMuxes = 4;
		this.channelsPerMux = 16;
	}

	/**
	 * Initialize all services
	 */
	async initialize() {
		console.log("Initializing Fuse Monitor Service...");

		try {
			// Initialize GPIO service
			GPIOService.initialize();
			console.log("GPIO service ready");

			// Initialize ADS1115 ADC
			await ADS1115Service.initialize();
			console.log("ADS1115 ADC ready");

			// Initialize CSV logger with proper headers
			const csvFilePath = process.env.CSV_FILE_PATH || "./data/fuse_data.csv";
			CSVLogger.initialize(csvFilePath);

			// Create CSV header if file is new
			await this.ensureCSVHeader();
			console.log("CSV logger ready");

			this.initialized = true;
			console.log("Fuse Monitor Service initialized successfully");
		} catch (error) {
			console.error("Failed to initialize Fuse Monitor Service:", error);
			throw error;
		}
	}

	/**
	 * Ensure CSV file has proper headers
	 */
	async ensureCSVHeader() {
		const headers = ["timestamp"];

		// Add headers for all 64 fuses
		for (let i = 1; i <= this.totalFuses; i++) {
			headers.push(`fuse ${i}`);
		}

		// Actually write the headers to the CSV file
		CSVLogger.writeHeaders(headers);
		console.log(`CSV headers written for ${this.totalFuses} fuses`);
	}

	/**
	 * Start monitoring loop
	 */
	startMonitoring() {
		if (!this.initialized) {
			throw new Error("Fuse Monitor Service not initialized");
		}

		if (this.monitoring) {
			console.log("Monitoring already started");
			return;
		}

		const interval = parseInt(process.env.DATA_COLLECTION_INTERVAL) || 5000;
		console.log(`Starting fuse monitoring every ${interval}ms`);

		this.monitoring = true;
		this.monitorInterval = setInterval(async () => {
			await this.collectAndLogData();
		}, interval);

		console.log("Fuse monitoring started");
	}

	/**
	 * Stop monitoring loop
	 */
	stopMonitoring() {
		if (this.monitorInterval) {
			clearInterval(this.monitorInterval);
			this.monitorInterval = null;
		}
		this.monitoring = false;
		console.log("Fuse monitoring stopped");
	}

	/**
	 * Collect data from all fuses and log to CSV
	 */
	async collectAndLogData() {
		const startTime = Date.now();
		const timestamp = new Date().toISOString().split(".")[0] + "Z"; // ISO format to seconds
		const fuseData = [];

		try {
			console.log("Starting data collection cycle...");

			// Collect data from all 4 multiplexers
			for (let muxIndex = 0; muxIndex < this.totalMuxes; muxIndex++) {
				console.log(`Reading mux ${muxIndex}...`);

				// Enable the current mux
				GPIOService.enableMux(muxIndex);
				await this.delay(1); // Wait 1ms after mux change

				// Read all 16 channels on this mux
				for (let channel = 0; channel < this.channelsPerMux; channel++) {
					try {
						// Set the channel on the mux
						GPIOService.setMuxChannel(channel);
						await this.delay(1); // Wait 1ms after channel change

						// Read voltage from the ADC channel corresponding to this mux
						const voltage = await ADS1115Service.readChannel(muxIndex);
						fuseData.push(voltage); // Format as "3.3"

						// Optional: Log detailed info for debugging
						const fuseNumber = muxIndex * this.channelsPerMux + channel + 1;
						console.log(
							`Fuse ${fuseNumber} (Mux ${muxIndex}, Ch ${channel}): ${voltage}V`,
						);
					} catch (error) {
						console.error(
							`Error reading mux ${muxIndex} channel ${channel}:`,
							error,
						);
						fuseData.push("0.0"); // Default value on error
					}
				}
			}

			// Disable all muxes after collection
			GPIOService.disableAllMux();

			// Log data to CSV
			const csvData = {
				timestamp: timestamp,
				values: fuseData,
			};

			await this.logToCSV(csvData);

			// Clean up variables to reduce memory usage
			fuseData.length = 0;

			const duration = Date.now() - startTime;
			console.log(`Data collection cycle completed in ${duration}ms`);
		} catch (error) {
			console.error("Error in data collection cycle:", error);

			// Ensure muxes are disabled even on error
			try {
				GPIOService.disableAllMux();
			} catch (gpioError) {
				console.error(
					"Error disabling muxes after collection error:",
					gpioError,
				);
			}
		}
	}

	/**
	 * Log data to CSV file
	 * @param {Object} csvData - Data object with timestamp and values
	 */
	async logToCSV(csvData) {
		try {
			// Create CSV row
			const row = [csvData.timestamp, ...csvData.values].join(",") + "\n";

			// Write to CSV using the logger's writeStream directly for efficiency
			const writeStream = CSVLogger.writeStream;
			if (writeStream) {
				await new Promise((resolve, reject) => {
					writeStream.write(row, (error) => {
						if (error) {
							reject(error);
						} else {
							resolve();
						}
					});
				});
			}

			console.log("Data logged to CSV successfully");
		} catch (error) {
			console.error("Error writing to CSV:", error);
		}
	}

	/**
	 * Test system functionality
	 */
	async testSystem() {
		if (!this.initialized) {
			throw new Error("Fuse Monitor Service not initialized");
		}

		console.log("Testing system functionality...");

		const results = {
			gpio: null,
			adc: null,
			testReadings: [],
		};

		try {
			// Test GPIO
			console.log("Testing GPIO pins...");
			results.gpio = await GPIOService.testPins();

			// Test ADC
			console.log("Testing ADC health...");
			results.adc = await ADS1115Service.isHealthy();

			// Test reading from first few channels
			console.log("Testing sample readings...");
			for (let mux = 0; mux < 2; mux++) {
				// Test first 2 muxes
				GPIOService.enableMux(mux);
				await this.delay(1);

				for (let channel = 0; channel < 4; channel++) {
					// Test first 4 channels
					GPIOService.setMuxChannel(channel);
					await this.delay(1);

					try {
						const voltage = await ADS1115Service.readChannel(mux);
						results.testReadings.push({
							mux: mux,
							channel: channel,
							voltage: voltage,
						});
					} catch (error) {
						results.testReadings.push({
							mux: mux,
							channel: channel,
							error: error.message,
						});
					}
				}
			}

			// Disable all muxes after test
			GPIOService.disableAllMux();

			console.log("System test completed");
			return results;
		} catch (error) {
			console.error("Error during system test:", error);
			results.error = error.message;
			return results;
		}
	}

	/**
	 * Get current system status
	 */
	getStatus() {
		return {
			initialized: this.initialized,
			monitoring: this.monitoring,
			totalFuses: this.totalFuses,
			gpio: GPIOService.getInfo(),
			adc: ADS1115Service.getInfo(),
			csv: CSVLogger.getFileStats(),
			currentMux: GPIOService.getEnabledMux(),
			currentChannel: GPIOService.getCurrentChannel(),
		};
	}

	/**
	 * Utility delay function
	 * @param {number} ms - Milliseconds to delay
	 */
	async delay(ms) {
		return new Promise((resolve) => setTimeout(resolve, ms));
	}

	/**
	 * Shutdown the service
	 */
	shutdown() {
		console.log("Shutting down Fuse Monitor Service...");

		// Stop monitoring
		this.stopMonitoring();

		// Shutdown GPIO
		GPIOService.shutdown();

		// Close CSV logger
		CSVLogger.close();

		this.initialized = false;
		console.log("Fuse Monitor Service shutdown complete");
	}

	/**
	 * Check if service is initialized
	 */
	isInitialized() {
		return this.initialized;
	}

	/**
	 * Check if monitoring is active
	 */
	isMonitoring() {
		return this.monitoring;
	}
}

// Export singleton instance
module.exports = new FuseMonitorService();

File: src/services/fuse_monitor_service.py
"""
Fuse Monitor Service
Coordinates GPIO, ADS1115 ADC, and CSV logging for 64-fuse monitoring system
Optimized for Raspberry Pi 1 Model B+ ARM6 architecture
"""

import os
import asyncio
import logging
from typing import Dict, Any, Optional
from datetime import datetime

from .gpio_service import GPIOService
from .ads1115_service import ADS1115Service
from .csv_logger import CSVLogger

logger = logging.getLogger(__name__)

class FuseMonitorService:
    """Main service coordinating fuse monitoring system"""
    
    def __init__(self):
        self.initialized = False
        self.monitoring = False
        self.monitor_task: Optional[asyncio.Task] = None
        
        # Hardware configuration
        self.total_fuses = 64  # 4 muxes × 16 channels each
        self.total_muxes = 4
        self.channels_per_mux = 16
        
        # Service instances
        self.gpio_service: Optional[GPIOService] = None
        self.ads1115_service: Optional[ADS1115Service] = None
        self.csv_logger: Optional[CSVLogger] = None
        
        # Monitoring configuration
        self.data_collection_interval = 5.0  # seconds
        self.current_mux = 0
        self.current_channel = 0
    
    async def initialize(self):
        """Initialize all services"""
        if self.initialized:
            logger.info("Fuse Monitor Service already initialized")
            return
        
        try:
            logger.info("Initializing Fuse Monitor Service...")
            
            # Initialize GPIO service
            self.gpio_service = GPIOService()
            await self.gpio_service.initialize()
            logger.info("GPIO service ready")
            
            # Initialize ADS1115 ADC
            self.ads1115_service = ADS1115Service()
            await self.ads1115_service.initialize()
            
            # Set appropriate gain for fuse monitoring (typically 0-5V)
            self.ads1115_service.set_gain('6.144V')  # +/-6.144V range for 0-5V fuse signals
            logger.info("ADS1115 ADC ready")
            
            # Initialize CSV logger
            csv_file_path = os.getenv('CSV_FILE_PATH', './data/fuse_data.csv')
            self.csv_logger = CSVLogger()
            await self.csv_logger.initialize(csv_file_path)
            
            # Write CSV headers for all 64 fuses
            await self._ensure_csv_headers()
            logger.info("CSV logger ready")
            
            # Get monitoring configuration
            self.data_collection_interval = float(os.getenv('DATA_COLLECTION_INTERVAL', 5.0))
            
            self.initialized = True
            logger.info("Fuse Monitor Service initialized successfully")
            logger.info(f"Monitoring configuration: {self.total_fuses} fuses, "
                       f"{self.data_collection_interval}s interval")
            
        except Exception as e:
            logger.error(f"Failed to initialize Fuse Monitor Service: {e}")
            await self.cleanup()
            raise
    
    async def _ensure_csv_headers(self):
        """Ensure CSV file has proper headers for all 64 fuses"""
        try:
            headers = ['timestamp']
            
            # Add headers for all 64 fuses
            for fuse_num in range(1, self.total_fuses + 1):
                headers.append(f'fuse {fuse_num}')
            
            await self.csv_logger.write_headers(headers)
            logger.info(f"CSV headers configured for {self.total_fuses} fuses")
            
        except Exception as e:
            logger.error(f"Failed to configure CSV headers: {e}")
            raise
    
    async def test_system(self) -> Dict[str, Any]:
        """
        Run comprehensive system test
        
        Returns:
            Dictionary with test results
        """
        if not self.initialized:
            raise RuntimeError("Fuse Monitor Service not initialized")
        
        test_results = {
            'gpio': False,
            'adc': False,
            'csv': False,
            'fuse_test': {},
            'error': None,
            'timestamp': datetime.now().isoformat()
        }
        
        try:
            logger.info("Running system test...")
            
            # Test GPIO functionality
            gpio_test = await self.gpio_service.test_gpio_pins()
            test_results['gpio'] = all(gpio_test.values())
            test_results['gpio_details'] = gpio_test
            
            # Test ADS1115 ADC
            adc_test = await self.ads1115_service.test_all_channels()
            test_results['adc'] = adc_test.get('connection', False)
            test_results['adc_details'] = adc_test
            
            # Test CSV logger
            csv_status = await self.csv_logger.get_status()
            test_results['csv'] = csv_status.get('initialized', False)
            test_results['csv_details'] = csv_status
            
            # Test a few sample fuses
            if test_results['gpio'] and test_results['adc']:
                test_fuses = [1, 17, 33, 49]  # One from each mux
                for fuse_num in test_fuses:
                    try:
                        voltage = await self._read_fuse(fuse_num)
                        test_results['fuse_test'][fuse_num] = {
                            'voltage': voltage,
                            'success': True
                        }
                    except Exception as e:
                        test_results['fuse_test'][fuse_num] = {
                            'error': str(e),
                            'success': False
                        }
                        logger.error(f"Fuse {fuse_num} test failed: {e}")
            
            # Overall test result
            test_results['overall'] = (test_results['gpio'] and 
                                     test_results['adc'] and 
                                     test_results['csv'])
            
            logger.info(f"System test complete. Overall result: {test_results['overall']}")
            return test_results
            
        except Exception as e:
            test_results['error'] = str(e)
            logger.error(f"System test failed: {e}")
            return test_results
    
    async def start_monitoring(self):
        """Start the monitoring loop"""
        if not self.initialized:
            raise RuntimeError("Fuse Monitor Service not initialized")
        
        if self.monitoring:
            logger.info("Monitoring already started")
            return
        
        try:
            logger.info(f"Starting fuse monitoring every {self.data_collection_interval}s")
            
            self.monitoring = True
            self.monitor_task = asyncio.create_task(self._monitoring_loop())
            
            logger.info("Fuse monitoring started successfully")
            
        except Exception as e:
            logger.error(f"Failed to start monitoring: {e}")
            self.monitoring = False
            raise
    
    async def stop_monitoring(self):
        """Stop the monitoring loop"""
        if not self.monitoring:
            return
        
        try:
            logger.info("Stopping fuse monitoring...")
            
            self.monitoring = False
            
            if self.monitor_task:
                self.monitor_task.cancel()
                try:
                    await self.monitor_task
                except asyncio.CancelledError:
                    pass
                self.monitor_task = None
            
            # Disable all multiplexers
            if self.gpio_service:
                await self.gpio_service.disable_all_mux()
            
            logger.info("Fuse monitoring stopped")
            
        except Exception as e:
            logger.error(f"Error stopping monitoring: {e}")
    
    async def _monitoring_loop(self):
        """Main monitoring loop"""
        logger.info("Starting monitoring loop...")
        
        try:
            while self.monitoring:
                start_time = asyncio.get_event_loop().time()
                
                # Collect data from all fuses
                fuse_data = await self._collect_all_fuse_data()
                
                # Log to CSV
                await self.csv_logger.log_fuse_readings(fuse_data)
                
                # Calculate actual collection time
                collection_time = asyncio.get_event_loop().time() - start_time
                logger.debug(f"Data collection completed in {collection_time:.2f}s")
                
                # Sleep for remaining interval time
                sleep_time = max(0, self.data_collection_interval - collection_time)
                if sleep_time > 0:
                    await asyncio.sleep(sleep_time)
                else:
                    logger.warning(f"Data collection took longer than interval "
                                 f"({collection_time:.2f}s > {self.data_collection_interval}s)")
                    
        except asyncio.CancelledError:
            logger.info("Monitoring loop cancelled")
            raise
        except Exception as e:
            logger.error(f"Monitoring loop error: {e}")
            self.monitoring = False
            raise
    
    async def _collect_all_fuse_data(self) -> Dict[int, float]:
        """
        Collect voltage readings from all 64 fuses
        
        Returns:
            Dictionary mapping fuse numbers to voltage readings
        """
        fuse_data = {}
        
        try:
            # Iterate through all muxes and channels
            for mux_index in range(self.total_muxes):
                for channel in range(self.channels_per_mux):
                    fuse_number = (mux_index * self.channels_per_mux) + channel + 1
                    
                    try:
                        voltage = await self._read_fuse(fuse_number)
                        fuse_data[fuse_number] = voltage
                        
                    except Exception as e:
                        logger.error(f"Failed to read fuse {fuse_number}: {e}")
                        fuse_data[fuse_number] = 0.0  # Use 0V for failed readings
            
            logger.debug(f"Collected data from {len(fuse_data)} fuses")
            return fuse_data
            
        except Exception as e:
            logger.error(f"Failed to collect fuse data: {e}")
            raise
    
    async def _read_fuse(self, fuse_number: int) -> float:
        """
        Read voltage from a specific fuse
        
        Args:
            fuse_number: Fuse number (1-64)
            
        Returns:
            Voltage reading in volts
        """
        if not 1 <= fuse_number <= self.total_fuses:
            raise ValueError(f"Fuse number must be 1-{self.total_fuses}")
        
        try:
            # Select the appropriate multiplexer and channel
            await self.gpio_service.select_fuse(fuse_number)
            
            # Small settling delay for multiplexer switching
            await asyncio.sleep(0.005)
            
            # Read from ADC channel 0 (all mux outputs go to AIN0)
            voltage = await self.ads1115_service.read_channel(0)
            
            # Update tracking variables
            fuse_index = fuse_number - 1
            self.current_mux = fuse_index // self.channels_per_mux
            self.current_channel = fuse_index % self.channels_per_mux
            
            return voltage
            
        except Exception as e:
            logger.error(f"Failed to read fuse {fuse_number}: {e}")
            raise
    
    async def get_status(self) -> Dict[str, Any]:
        """
        Get current system status
        
        Returns:
            Status dictionary
        """
        status = {
            'initialized': self.initialized,
            'monitoring': self.monitoring,
            'current_mux': self.current_mux,
            'current_channel': self.current_channel,
            'data_collection_interval': self.data_collection_interval,
            'total_fuses': self.total_fuses
        }
        
        # Add service statuses
        if self.gpio_service:
            status['gpio'] = await self.gpio_service.get_status()
        
        if self.ads1115_service:
            status['adc'] = await self.ads1115_service.get_status()
        
        if self.csv_logger:
            status['csv'] = await self.csv_logger.get_status()
        
        return status
    
    async def cleanup(self):
        """Cleanup all services"""
        try:
            logger.info("Cleaning up Fuse Monitor Service...")
            
            # Stop monitoring
            await self.stop_monitoring()
            
            # Cleanup services
            if self.csv_logger:
                await self.csv_logger.cleanup()
            
            if self.ads1115_service:
                await self.ads1115_service.cleanup()
            
            if self.gpio_service:
                await self.gpio_service.cleanup()
            
            self.initialized = False
            logger.info("Fuse Monitor Service cleanup complete")
            
        except Exception as e:
            logger.error(f"Error during Fuse Monitor Service cleanup: {e}")
    
    async def shutdown(self):
        """Shutdown Fuse Monitor Service"""
        await self.cleanup()
    
    def is_initialized(self) -> bool:
        """Check if Fuse Monitor Service is initialized"""
        return self.initialized
